sycl/include/sycl/nd_item.hpp

```diff
@@ -38,6 +41,492 @@ namespace ext::oneapi::experimental {
+    range<Dimensions> Extent = get_local_range();
+    if (1 == Dimensions) {
+      LinId = Index[0];
+    } else if (2 == Dimensions) {
+      LinId = Index[0] * Extent[1] + Index[1];
+    } else {
+      LinId =
+          Index[0] * Extent[1] * Extent[2] + Index[1] * Extent[2] + Index[2];
+    }
+    __SYCL_ASSUME_INT(LinId);
+    return LinId;
+  }
+
+  group<Dimensions> get_group() const {
+    // TODO: ideally Group object should be stateless and have a contructor with
+    // no arguments.
+    return detail::Builder::createGroup(get_global_range(), get_local_range(),
+                                        get_group_range(), get_group_id());
+  }
+
+  sub_group get_sub_group() const { return sub_group(); }
+
+  size_t __SYCL_ALWAYS_INLINE get_group(int Dimension) const {
+    size_t Id = get_group_id()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_linear_id() const {
+    size_t LinId = 0;
+    id<Dimensions> Index = get_group_id();
+    range<Dimensions> Extent = get_group_range();
+    if (1 == Dimensions) {
+      LinId = Index[0];
+    } else if (2 == Dimensions) {
+      LinId = Index[0] * Extent[1] + Index[1];
+    } else {
+      LinId =
+          Index[0] * Extent[1] * Extent[2] + Index[1] * Extent[2] + Index[2];
+    }
+    __SYCL_ASSUME_INT(LinId);
+    return LinId;
+  }
+
+  range<Dimensions> get_group_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initNumWorkgroups<Dimensions, range<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  size_t __SYCL_ALWAYS_INLINE get_group_range(int Dimension) const {
+    size_t Range = get_group_range()[Dimension];
+    __SYCL_ASSUME_INT(Range);
+    return Range;
+  }
+
+  range<Dimensions> get_global_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalSize<Dimensions, range<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  size_t get_global_range(int Dimension) const {
+    size_t Val = get_global_range()[Dimension];
+    __SYCL_ASSUME_INT(Val);
+    return Val;
+  }
+
+  range<Dimensions> get_local_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initWorkgroupSize<Dimensions, range<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  size_t get_local_range(int Dimension) const {
+    size_t Id = get_local_range()[Dimension];
+    __SYCL_ASSUME_INT(Id);
+    return Id;
+  }
+
+  __SYCL2020_DEPRECATED("offsets are deprecated in SYCL 2020")
+  id<Dimensions> get_offset() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return __spirv::initGlobalOffset<Dimensions, id<Dimensions>>();
+#else
+    assert(false && "nd_item methods can't be used on the host!");
+    return {};
+#endif
+  }
+
+  nd_range<Dimensions> get_nd_range() const { return nd_range<Dimensions>(); }
```

Comment:
Previous implementation passes arguments to `nd_range`. Don't we need them here?

Reply:
Nice catch. I think I just forgot to implement this method.
The scary thing is that pre-commit passed! We have 310 end-to-end tests using `nd_item` and none of them test this method.

Reply:
Make sense to me. I've filed https://github.com/intel/llvm/issues/12563 to follow-up on that.